From 3f16f4a34eeb22d70afe1bebdd2cfcded179f22a Mon Sep 17 00:00:00 2001
From: Jeremy Cline <jcline@redhat.com>
Date: Tue, 4 Feb 2020 15:31:19 -0500
Subject: [PATCH] Include kvm_asm.h and kvm_arm.h in kvm/arm/trace.h

commit 0e20f5e25556 maybe broke this.

Signed-off-by: Jeremy Cline <jcline@redhat.com>
---
 virt/kvm/arm/trace.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/virt/kvm/arm/trace.h b/virt/kvm/arm/trace.h
index 204d210d01c2..1c88e60a6b00 100644
--- a/virt/kvm/arm/trace.h
+++ b/virt/kvm/arm/trace.h
@@ -4,6 +4,8 @@
 
 #include <kvm/arm_arch_timer.h>
 #include <linux/tracepoint.h>
+#include <asm/kvm_asm.h>
+#include <asm/kvm_arm.h>
 
 #undef TRACE_SYSTEM
 #define TRACE_SYSTEM kvm
-- 
2.24.1

