diff -up linux-2.6.34.noarch/arch/x86/Makefile.orig linux-2.6.34.noarch/arch/x86/Makefile
--- linux-2.6.34.noarch/arch/x86/Makefile.orig	2010-07-01 13:33:21.859627499 -0400
+++ linux-2.6.34.noarch/arch/x86/Makefile	2010-07-01 13:36:26.751576450 -0400
@@ -81,6 +81,7 @@ ifdef CONFIG_CC_STACKPROTECTOR
                 $(warning stack protector enabled but no compiler support)
         endif
 endif
+KBUILD_CFLAGS += -fno-inline-functions-called-once
 
 # Don't unroll struct assignments with kmemcheck enabled
 ifeq ($(CONFIG_KMEMCHECK),y)
