diff --git a/arch/x86/pci/acpi.c b/arch/x86/pci/acpi.c
index 404f21a..2704a9c 100644
--- a/arch/x86/pci/acpi.c
+++ b/arch/x86/pci/acpi.c
@@ -24,6 +24,12 @@ static int __init set_use_crs(const struct dmi_system_id *id)
 	return 0;
 }
 
+static int __init set_nouse_crs(const struct dmi_system_id *id)
+{
+	pci_use_crs = false;
+	return 0;
+}
+
 static const struct dmi_system_id pci_use_crs_table[] __initconst = {
 	/* http://bugzilla.kernel.org/show_bug.cgi?id=14183 */
 	{
@@ -54,6 +60,18 @@ static const struct dmi_system_id pci_use_crs_table[] __initconst = {
 			DMI_MATCH(DMI_BIOS_VENDOR, "American Megatrends Inc."),
 		},
 	},
+
+	/* Now for the blacklist.. */
+
+	/* https://bugzilla.redhat.com/show_bug.cgi?id=770308 */
+	{
+		.callback = set_nouse_crs,
+		.ident = "Dell Studio 1536",
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "Dell Inc."),
+			DMI_MATCH(DMI_BOARD_NAME, "0M273C"),
+		},
+	},
 	{}
 };
 
--- linux-3.1.noarch/arch/x86/pci/acpi.c~	2011-12-30 14:42:43.495717907 -0500
+++ linux-3.1.noarch/arch/x86/pci/acpi.c	2011-12-30 14:45:09.296198024 -0500
@@ -72,6 +72,16 @@ static const struct dmi_system_id pci_us
 			DMI_MATCH(DMI_BOARD_NAME, "0M273C"),
 		},
 	},
+	/* https://bugzilla.redhat.com/show_bug.cgi?id=769657 */
+	{
+		.callback = set_nouse_crs,
+		.ident = "Thinkpad SL510",
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_BOARD_NAME, "2847DFG"),
+			DMI_MATCH(DMI_BIOS_VERSION, "6JET85WW (1.43 )"),
+		},
+	},
 	{}
 };
 
--- linux-3.1.noarch/arch/x86/pci/acpi.c~	2012-01-04 11:19:36.783664477 -0500
+++ linux-3.1.noarch/arch/x86/pci/acpi.c	2012-01-04 11:23:42.403666272 -0500
@@ -75,6 +75,16 @@ static const struct dmi_system_id pci_us
 	/* https://bugzilla.redhat.com/show_bug.cgi?id=769657 */
 	{
 		.callback = set_nouse_crs,
+		.ident = "Dell Studio 1557",
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "Dell Inc."),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Studio 1557"),
+			DMI_MATCH(DMI_BIOS_VERSION, "A09"),
+		},
+	},
+	/* https://bugzilla.redhat.com/show_bug.cgi?id=769657 */
+	{
+		.callback = set_nouse_crs,
 		.ident = "Thinkpad SL510",
 		.matches = {
 			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
