revert c055f5b2614b4f758ae6cc86733f31fa4c2c5844 from 2.6.38.6

--- b/drivers/scsi/scsi_lib.c
+++ a/drivers/scsi/scsi_lib.c
@@ -400,15 +400,10 @@
 static void scsi_run_queue(struct request_queue *q)
 {
 	struct scsi_device *sdev = q->queuedata;
+	struct Scsi_Host *shost = sdev->host;
-	struct Scsi_Host *shost;
 	LIST_HEAD(starved_list);
 	unsigned long flags;
 
-	/* if the device is dead, sdev will be NULL, so no queue to run */
-	if (!sdev)
-		return;
-
-	shost = sdev->host;
 	if (scsi_target(sdev)->single_lun)
 		scsi_single_lun_run(sdev);
 
