diff --git a/drivers/net/bonding/bond_main.c b/drivers/net/bonding/bond_main.c
index 0075514..136963b 100644
--- a/drivers/net/bonding/bond_main.c
+++ b/drivers/net/bonding/bond_main.c
@@ -3285,7 +3285,7 @@ static void bond_create_proc_entry(struct bonding *bond)
 	struct net_device *bond_dev = bond->dev;
 	struct bond_net *bn = net_generic(dev_net(bond_dev), bond_net_id);
 
-	if (bn->proc_dir) {
+	if (bn->proc_dir && !bond->proc_entry) {
 		bond->proc_entry = proc_create_data(bond_dev->name,
 						    S_IRUGO, bn->proc_dir,
 						    &bond_info_fops, bond);
