diff -ur linux-2.6.33.noarch.bak/drivers/platform/x86/thinkpad_acpi.c linux-2.6.33.noarch/drivers/platform/x86/thinkpad_acpi.c
--- linux-2.6.33.noarch.bak/drivers/platform/x86/thinkpad_acpi.c	2010-05-17 15:56:10.864200185 -0400
+++ linux-2.6.33.noarch/drivers/platform/x86/thinkpad_acpi.c	2010-05-17 16:12:20.232074844 -0400
@@ -3396,7 +3396,7 @@
 		/* update bright_acpimode... */
 		tpacpi_check_std_acpi_brightness_support();
 
-	if (tp_features.bright_acpimode && acpi_video_backlight_support()) {
+	if (acpi_video_backlight_support()) {
 		printk(TPACPI_INFO
 		       "This ThinkPad has standard ACPI backlight "
 		       "brightness control, supported by the ACPI "
@@ -6187,26 +6187,24 @@
 	 * going to publish a backlight interface
 	 */
 	b = tpacpi_check_std_acpi_brightness_support();
-	if (b > 0) {
 
-		if (acpi_video_backlight_support()) {
-			if (brightness_enable > 1) {
-				printk(TPACPI_NOTICE
-				       "Standard ACPI backlight interface "
-				       "available, not loading native one.\n");
-				return 1;
-			} else if (brightness_enable == 1) {
-				printk(TPACPI_NOTICE
-				       "Backlight control force enabled, even if standard "
-				       "ACPI backlight interface is available\n");
-			}
-		} else {
-			if (brightness_enable > 1) {
-				printk(TPACPI_NOTICE
-				       "Standard ACPI backlight interface not "
-				       "available, thinkpad_acpi native "
-				       "brightness control enabled\n");
-			}
+	if (acpi_video_backlight_support()) {
+		if (brightness_enable > 1) {
+			printk(TPACPI_NOTICE
+			       "Standard ACPI backlight interface "
+			       "available, not loading native one.\n");
+			return 1;
+		} else if (brightness_enable == 1) {
+			printk(TPACPI_NOTICE
+			       "Backlight control force enabled, even if standard "
+			       "ACPI backlight interface is available\n");
+		}
+	} else {
+		if (brightness_enable > 1) {
+			printk(TPACPI_NOTICE
+			       "Standard ACPI backlight interface not "
+			       "available, thinkpad_acpi native "
+			       "brightness control enabled\n");
 		}
 	}
 
